core/src/main/java/org/wildfly/extras/creaper/core/online/HttpModelControllerClient.java

```diff
@@ -124,33 +135,43 @@ public OperationResponse executeOperation(Operation operation, OperationMessageH
         return OperationResponse.Factory.createSimple(execute(operation));
     }
 
-    /**
-     * <b>Not supported!</b>
-     */
     @Override
     public AsyncFuture<ModelNode> executeAsync(ModelNode modelNode, OperationMessageHandler handler) {
-        throw new UnsupportedOperationException("Asynchronous execution is not supported by " + getClass().getName());
+        ExecutorService executor = obtainExecutorService();
+        Future<ModelNode> submit = executor.submit(new ModelNodeFromModelNodeCallable(modelNode));
+        ModelNode result = getFutureResult(submit);
+        return new HttpModelControllerClientAsyncFutureTask(executor, result);
     }
 
-    /**
-     * <b>Not supported!</b>
-     */
     @Override
     public AsyncFuture<ModelNode> executeAsync(Operation operation, OperationMessageHandler handler) {
-        throw new UnsupportedOperationException("Asynchronous execution is not supported by " + getClass().getName());
+        ExecutorService executor = obtainExecutorService();
+        Future<ModelNode> submit = executor.submit(new ModelNodeFromOperationCallable(operation));
+        ModelNode result = getFutureResult(submit);
+        return new HttpModelControllerClientAsyncFutureTask(executor, result);
     }
 
-    /**
-     * <b>Not supported!</b>
-     */
     @Override
     public AsyncFuture<OperationResponse> executeOperationAsync(Operation operation, OperationMessageHandler handler) {
-        throw new UnsupportedOperationException("Asynchronous execution is not supported by " + getClass().getName());
+        ExecutorService executor = obtainExecutorService();
+        Future<OperationResponse> submit = executor.submit(new OperationResponseFromOperationCallable(operation));
+        OperationResponse result = getFutureResult(submit);
+        return new HttpModelControllerClientAsyncFutureTask(executor, result);
     }
 
     @Override
     public void close() throws IOException {
         client.close();
+        if (executorService != null) {
+            executorService.shutdown();
+        }
+    }
+
+    private ExecutorService obtainExecutorService() {
```

Comment:
I don't think the executor service needs to be created lazily; if the client is created, it's extremely likely to be used :-) But I'll leave it to you -- this is good too.

Reply:
I created it lazily because older version than WildFly 12 still uses `execute` method instead of `executeAsync`. I believe it is better to created ExecutorService only on demand.

Reply:
Ha, very much true! Makes sense.